### DISCUSSION TITLE
Media player control feedback

### DISCUSSION THREAD

**Discussion by edwardtfn** [Blackymas/NSPanel_HA_Blueprint#1030] [General]
Support to Media Player is available on `dev` and we would appreciate if you can give some early feedback, so we can adjust it before going to a wider public.

You will be able to add a Media Player entity to any of the buttons pages and with a long press you will see the media player page.

Feel free to give any kind of feedback on the user interface, behavior and report any bug. I cannot make big promises for the first release, but it would be nice to hear new ideas anyways.

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/94725493/bf8b0a3d-a121-4bdd-81dc-22515405f33f)

**Comment by moustic999** [comment]
Damn it looks goods !

**Comment by Doughboy68** [comment]
Latest version today no longer displays the text info.
Also does not update the the play/pause button like previous version when exiting/going back into page.
I realize this is a work in progress.
There may be enough room on screen to use the next size font... transport controls seem a bit small (but then again I think climate is also too small).

**Reply by edwardtfn** [reply]
It should be working.
Do you have anything in your logs (HA and ESPHome)?
Also, what is the value for suported_features on this entity?

**Reply by Doughboy68** [reply]
The song name/title was showing before I did update. But now it is not even displaying any text.
It never did update live though.

Here is info for my homepod mini.

volume_level: 0.23
media_content_type: music
media_duration: 508
media_position: 7
media_position_updated_at: '2023-09-13T01:15:17.691536+00:00'
media_title: If Eternity Should Fail
media_artist: Iron Maiden
media_album_name: The Book of Souls
app_id: com.apple.Music
app_name: AirMusic
shuffle: true
repeat: 'off'
entity_picture: >-
  /api/media_player_proxy/media_player.homepod_mini_family_room?token=(removed)
friendly_name: HomePod Mini Family Room
supported_features: 448439

And couple errors in log I noticed after playing for a second then stopping..

Error while executing automation automation.nspanel_05_configuration: invalid boolean value None for dictionary value @ data['is_volume_muted']
11:16:39 PM – (ERROR) Automation
NSPanel 05 Configuration: Main choices: NSPanel event: Choose at step 2: Page changed: Choose at step 1: Media player settings page: Refresh page in a loop: Error executing script. Invalid data for call_service at pos 4: invalid boolean value None for dictionary value @ data['is_volume_muted']
11:16:39 PM – (ERROR) Automation - message first occurred at 11:16:39 PM and shows up 5 times

**Reply by edwardtfn** [reply]
I cannot duplicate this, but I've just pushed a fix for when an expected value is not provided (`is_volume_muted` in this case), preventing it to crash.
Please let me knot if it was fixed.
You have to update the Blueprint only.

**Reply by Doughboy68** [reply]
This fixed it. 

Seems to take 10 seconds to update after pressing a button on nspanel eg. play/pause. But works now! I see why because you loop it with a delay of 10 seconds so this explains. Is it possible to update instantly when a button is pressed on the panel instead of waiting up to 10 seconds?

I didnt realize maybe I had volume muted testing recently since I was testing panel downstairs and homepod was in a different room... late at night didn't want to bother anyone!

**Reply by edwardtfn** [reply]
> Seems to take 10 seconds to update after pressing a button on nspanel eg. play/pause. But works now! I see why because you loop it with a delay of 10 seconds so this explains. Is it possible to update instantly when a button is pressed on the panel instead of waiting up to 10 seconds?

I don't have a wait of 10sec, what I have is a loop to refresh the whole page every 10sec as the info might have changed.
If you see in that loop I'm calling for `homeassistant.update_entity` and for some reason those `media_player` entities are very slow to refresh it's states.
But what you are saying is about a reaction time for a command, and for this there's no delay at all. You can listen for events from `esphome.nspanel_ha_blueprint' that the command comes right after you press it. It's probably your `media_player' system taking long to react to that. I have the same issue with my Echo devices.
Do you see this delay with all the commands or with a specific command? On my tests here I feel that some commands are faster than others.
Do you see delays when you call the services from Home Assistant directly?

**Reply by Doughboy68** [reply]
Ok to clarify,  When I press play button on NSPanel, there is no delay of the command being processed and it starts playing almost immediately on the actual device, and Home Assistant media card for that device also updates almost immediately and the play icon will change to pause and track is playing. But, the NSPanel's media page does not update the icon from play to a pause icon until around 10 seconds later. I also see this same issue in other commands like skip next/previous track it does not update the text of the artist/song name for around 10 seconds. This makes sense if you are updating the information every 10 seconds. But what I am asking (not looking at code, just speculating at a higher level here) is do you call the update_entity after a button is pressed so it updates the NSPanel right away? If you do, maybe this call is just a bit too soon before HA receives the command, and it is updating old information, then the 10 second loop ends up updating it to correct values again. Does make sense? If this is the way it is working, maybe putting a small delay like 2 seconds before calling update_entity after a button press.

**Reply by edwardtfn** [reply]
For all the other domains, when you click in the button, the event opens one instance of the blueprint which will call the service and then end itself.
When the entity being controlled change it's state, it generates another event on HA, which will trigger the blueprint again, but this time it will go thru the track to update the screen.
This is how is the reaction for all the other domains.

For `media_player`, however, that takes an eternity to the entity generates its event with the change, so I've added all that logic for calling a refresh. Again, when a refresh happens, it should immediately trigger another instance of the blueprint to update it.

Another question...
If you could monitor at Developer Tools > State, how long it takes since the device changes it states to have that reflected in the Developer Tools > State page?

**Comment by edwardtfn** [comment]
> There may be enough room on screen to use the next size font... transport controls seem a bit small (but then again I think climate is also too small).

Is it hard to see or hard to click?
I'm asking this as changing the font size doesn't change the clickable area. These are two independent attributes.

I'm afraid increasing the font will affect all the other pages where that same font is used. Might not be an issue, but requires review. 
And will increase the tft size.

**Reply by Doughboy68** [reply]
On a smaller panel seems a bit small. Guess that is just the way it is.
Looking at HMI the actual spacing seems like not much room to make larger.

Is the stop button necessary? The play/pause button seem to do the same thing as pressing stop. I do not see a stop button in HA player, just play/pause. But it could be just my Homepod player does not have that separate stop button.

**Reply by edwardtfn** [reply]
> Is the stop button necessary?

I don't see those necessary for my case, but I don't know about the others.
The challenge with `media_player` from the beginning is the huge amount of possibilities with all the different `supported_features`:

https://developers.home-assistant.io/docs/core/entity/media-player/#supported-features

All the different devices out there can have any combination of those features.

**Reply by edwardtfn** [reply]
>  I thought there was a slightly larger font you are using for fan speed control that is already in the tft.

You're right.
I've just realized I've used font size 32. We do have size 48 available in the tft.
I will take a look on how to fit a bigger font there.

**Comment by Doughboy68** [comment]
After all the fixes/updates it seems to work good except the volume can be a bit jumpy and inconsistent. What happens is if sliding the volume slider and it does a refresh it jumps back to the last value and does not change the volume when let go. It is hard to explain and best if you try and slide volume around you may see what I mean. Sometimes this even happens when pressing a location on the slider without sliding. Is it possible to give preference to the volume value from the tft when let go, regardless what the refreshed value is set to?

**Reply by edwardtfn** [reply]
I will take a look at that.

> Is it possible to give preference to the volume value from the tft when let go, regardless what the refreshed value is set to?

We had some similar problem with Climate slider, where moving the slider was triggering hundreds of calls to the climate entity and then each one of those triggered a refresh to the page with the volume in a different place. In the end we solved with a timer, so we don't call HA to change it unless it stopped changing for a second.
I'm not sure if this was implemente with volume, I will take a look. It might be a way to improve.

**Reply by edwardtfn** [reply]
I just saw this timer is also used for volume, so it takes a second from when you change the volume to that be send to Home Assistant.

The problem again is with the slow response from the media player entity and the refresh process. I get a new value from that entity before the new value was sent, so the slider gets updated when I still moving it.

I will play a bit with that.

**Reply by edwardtfn** [reply]
So, I've just pushed a new TFT with 500ms for the volume wait (it was 1s before).
I've also made some changes to the UI (increased icons and moved power button on US model).
Please let me know your thoughts.

**Reply by edwardtfn** [reply]
Honestly, I liked more the smaller icons. And couldn't see an improve on the jumpy volume control. :(

**Comment by Doughboy68** [comment]
Ya the jumpy volume is maybe a bit better but often it doesn't update where it was set and jumps back. Probably fine since one is probably not going to be constantly changing volume. Another option would be to increment/decrement the volume by 10% instead of 1% with the plus/minus buttons (or have another advanced setting to allow modification). This would still jump but may not be as bad.

 And I agree, the small icons are probably nicer because of the spacing and fit better with the rest of the UI.
I notice in ESPHome log that there is a touchevent whenever touching the panel... could this be detected to disable the refresh, or maybe would be possible to disable send a command to esphome to disable this refresh when panel pressed and re-enable the refresh when released?

**Reply by edwardtfn** [reply]
I've made a small change in the volume control. It still not 100%, but I think it is much better than before (good enough?).
I've also removed button "Stop" and moved button "Power" on the US model, and with this increased the clickable area for buttons.
Could you please give me some feedback?

**Reply by Doughboy68** [reply]
Yes it seems better now. The odd occasion where the new volume doesn't "stick" but it is acceptable.
The layout also looks good.
One issue is the playback time position bar and time does not update... usually alternates between 0:01 and 0:02. Exiting page and going back in sometimes advances to correct position but then it doesn't update anymore.

**Reply by edwardtfn** [reply]
You are using `nspanel_us.tft`, right?
I will install that and try. It's probably something specific to that file as it works fine with `nspanel_us_land.tft` (which is a copy of `nspanel_eu.tft`).

**Reply by Doughboy68** [reply]
I will try the us_land in meantime also to see if it works to rule out something else.

**Reply by Doughboy68** [reply]
I tried and it does the same thing. I recompiled esphome and updated blueprint just in case.

It seems to reset back after the update interval. I have mine set lower to 2500ms so I notice. If setting back to 10000ms it appears that it is working then jumps back 10 seconds so maybe isn't noticeable.

**Reply by edwardtfn** [reply]
What if you go to Developer Tools > State, then look for the attributes on the Media Player entities, there is one name something like `media_position`... Is that being updated while the music is playing and displayed in the panel?

**Reply by Doughboy68** [reply]
Actually it seems to be looping the actual player itself according to what I see in HA... I usually test it remotely so will go into same room as a player and see what is going on.

**Reply by edwardtfn** [reply]
Maybe it didn't liked all the refresh requests... 😆

**Reply by Doughboy68** [reply]
I tried on a complete different panel and homepod and the time loops for duration of update interval while nspanel is playing, but the music plays fine... eg. if play position is currently at 1:30 then it will loop around 1:40 and go back to 1:30 (if update interval is set to 10000ms). If I have mine at 2500ms it loops every  2 sec.

**Reply by edwardtfn** [reply]
I've just pushed an untested fix for this. Could you please try it?
Blueprint only.

**Reply by edwardtfn** [reply]
Basically, the blueprint gets that position from your entity and send that to the panel, where a timer keeps the time moving every second (running directly on Nextion display). Probably after the refresh interval you are getting the same original value, then it restarts counting from that point.
I've added a correction where it uses another attribute to get from when that info is and adjust it to current time before sending to the panel.

**Reply by Doughboy68** [reply]
updated blueprint and no updates at all on the nspanel now.

**Reply by edwardtfn** [reply]
Any error messages on HA logs?

**Reply by Doughboy68** [reply]
Didn't notice, accidentally flashed main tft when going back to us portrait, waiting to flash proper one now and will look again.

**Reply by edwardtfn** [reply]
I found the bug.

```
Logger: homeassistant.components.automation.office_workstation_panel_configuration
Source: helpers/script.py:420
Integration: Automation (documentation, issues)
First occurred: 03:26:23 (2 occurrences)
Last logged: 03:26:27

Office - Workstation - Panel - Configuration: Main choices: NSPanel event: Choose at step 2: Page changed: Choose at step 1: Media player settings page: Refresh page in a loop: Error executing script. Error for call_service at pos 4: Error rendering data template: UndefinedError: 'dict object' has no attribute 'position'
```

Fixed!
Blueprint only.

**Reply by Doughboy68** [reply]
working again, but the time is still strange. The nspanel seems to put a different playback time and sends this to HA but it keeps playing on my device correctly. If I exit the nspanel while playing, let it play a bit I can see the time increasing in HA... then when I launch nspanel player again, it puts it way back to the last nspanel time. Its only when stopped/paused from nspanel it seems to reset this.

**Reply by edwardtfn** [reply]
The panel isn't sending anything to HA.
The Blueprint calls `homeassistant.update_entity` on the loop interval you set.

Try to close the Media Player page on your panel, then go to HA, on Developer Tools > Services and call that service against your media_player entity while playing... What happens to the music and to the attributes?

**Reply by Doughboy68** [reply]
It appears this may be limitation of Apple Homepod not sending updated media_position to Home Assistant while playing. Refreshing the states of my homepod entity, it does not update the position in states.

I think what I see happening is particular media player card I have is estimating the playback time itself within the card itself using its own timer, and a pause/play elsewhere is causing it to jump around.
Note that the media_position_updated_at state below gets updated every 2.5 seconds when nspanel page is open (my current update interval is 2500ms). This is why I notice it jumping around on HA card when nspanel media player page is open.

The states playback position for apple homepod only seems to update when initially playing or pausing the player. Maybeother players update the media_position more often.

I will look into this and see if this is expected with Apple Homepods. If that is the case, then maybe disable that position/timer bar and the time media_position for apple devices..

friendly_name: HomePod Family Room
supported_features: 448439
volume_level: 0.04
media_content_type: music
media_duration: 242
media_position: 208
media_position_updated_at: '2023-09-18T02:47:47.659465+00:00'
media_title: Reaping Stone (Remixed & Remastered)
media_artist: Iced Earth
media_album_name: Something Wicked This Way Comes
app_id: com.apple.Music
app_name: AirMusic
shuffle: true
repeat: 'off'
entity_picture: >-

**Reply by edwardtfn** [reply]
> Note that the media_position_updated_at state below gets updated every 2.5 seconds when nspanel page is open (my current update interval is 2500ms). This is why I notice it jumping around on HA card when nspanel media player page is open.

🤔
With `media_position_updated_at` being updated but `media_position` static it's impossible to calculate the time. 😩
Now I have to find a way to detect when this is happening so I can disable the display of that progress bar, which is a shame.

By the way, is the media position updated when a music ends and snother one starts?

**Reply by Doughboy68** [reply]
It updates to 1 second I think when new song is played.

**Reply by edwardtfn** [reply]
Ok, so maybe the only thing I have is the first timestamp when a song start... I will see what I can do with that.

**Comment by Doughboy68** [comment]
https://community.home-assistant.io/t/no-media-position-update/200406

I even tried homeassistant.update_entity like you suggested earlier and it does not update the apple homepod's media_position but it does update the media_last_updated at so something is wrong on my end. Probably something wrong with apple players. Will need others to test to see if this is something on my system.

Edit, in this thread others are saying that the service call does not update the media_position. There seem to be other solutions in there which require templates but look overly complicated for something that really is not a big deal.

https://community.home-assistant.io/t/calculating-current-media-position-for-use-in-automation/240087/6

This explains why the built-in cards in HA do not display the time and it was only one card I found in HACS that did.
I checked how lovelace ui does it, and appears they just do not display any times or progress bar.

I think that the best solution is to disable the time completely from the page, or make it an option for those that have compatible players.

**Reply by edwardtfn** [reply]
I'm trying a slightly different approach. Could you please update Blueprint and ESPHome and let me know if you see any improvement?
Now the time will be only estimated if your HA isn't updating it, so we might have a couple of seconds delay on the time position, but at least will give an idea on where you are in the music.

**Comment by Doughboy68** [comment]
On a different subject than the media_position, I notice there should be two additional buttons... a shuffle button and repeat mode.

**Reply by edwardtfn** [reply]
Version 4.1... 😉

**Reply by quenthal** [reply]
Excellent :).

Also few customizable buttons for choosing source /urls for mediaplayer; ie. specific playlists, radios etc. So you could choose music from panel. Maybe we could even point to these new cool action template sensors in 2023.09 somehow for cool automations...

Also something to think about is player groups, ie. only for room we are in, or all players in the house. But this is not that high priority.

**Reply by edwardtfn** [reply]
I don't wanna add new functionalities before the release, so I will now focus on fixing bugs with the functionalities already there.
It would be great if you can add new "feature request" in another topic (as a new issue),so we can keep track of those.
🙂

**Comment by Doughboy68** [comment]
I tried before went to work and does update the time but if exit and go back in it only uses the last known time (media postion). Better looking visually than jumping around I suppose. Most probably wouldn't notice.

I have to correct a previous observation I made previously. I did notice that the built-in Media Player card in HA does in fact have a progress bar at the bottom of the card, and it seems to keep track of the position somehow internally. I am not sure if cards are capable of running/updating in background but I didn't have time to experiment... eg was going to restart HA this morning while music is playing to try to determine if it "recovers" the correct time or it is using an internal counter. If it using an internal counter I would expect the displayed time to lag by the duration it took HA to restart. Maybe I can do some more research instead knowing more about this now.

Maybe it is possible to make a template in HA that monitors status of a media player entity and keeps a better approximation of the time taking into account pausing etc.. I think that was discussed in the threads I linked to. But NSPanel would have to be aware of this.

Also I did notice that built in HA increments volume by 5% when using the + - keys. This may work better on nspanel instead of 1% - what is your opinion?

**Reply by edwardtfn** [reply]
I have some devices where 5% is a lot, while in others you almost cannot notice any change. As the idea for the +/- is to make a fine adjustment (which is quite hard in the slider), I don't wanna hard code 5%.
Maybe it could be another input in the blueprint settings, so each user can choose (although I'm sure soon they will request this per device 😆 ).

**Reply by edwardtfn** [reply]
I wanna go for a release as it is, then we improve later with more feedback.

**Reply by Doughboy68** [reply]
Yes, I agree, would just prepare for release and work on for 4.1. Media player is a bonus anyways and you were not sure if was going to make it in!!!

**Comment by edwardtfn** [comment]
It's out for `beta` now: #1046

**Comment by edwardtfn** [comment]
I'm closing this discussion as we have released v4.0 (#1057).
Please add a feature request for the missing things.

And thanks a lot for your support.
